### DISCUSSION TITLE
Opentelemetry to export traces to Database

### DISCUSSION THREAD

**Discussion by Yahire3388** [open-telemetry/opentelemetry-java-instrumentation#5573] [Q&A]
Hi,

I am using opentelemetry java agent to trace web application and zipkin as a collector. Instead of exporting the data to zipkin, can we directly store the data in database like Postgres, mongo or cassandra?

Thanks in advance :)

**Comment by mateuszrzeszutek** [comment]
Hey @Yahire3388 ,
This could be achieved by implementing a custom [`SpanExporter`](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/SpanExporter.java) (and wiring it using a [`ConfigurableSpanExporterProvider`](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/traces/ConfigurableSpanExporterProvider.java)) in a custom [extension jar](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension). You could probably start with copying the extension project and trying it out - note that the extension jar is separate from your main application, and if you want to use a database you have to keep in mind that the extension will need to setup and manage connections  by itself.

**Reply by Yahire3388** [reply]
Hi @[mateuszrzeszutek](https://github.com/mateuszrzeszutek)

I wanted to proceed with auto instrumentation only. Also, I looked into the github repository to see if i can setup the code and debug it to add my code and create jar out of it, but i am not able to understand the structure of the code as i can see some gradle files, kotlin files. used following github project url : https://github.com/open-telemetry/opentelemetry-java-instrumentation.

Can you please help to setup the project in my local or which source code is used to create the jar, or provide with a custom jar which will export the traces directly to the database. This can be a new idea to save the traces to database directly eliminating the need of collector.

**Reply by trask** [reply]
> provide with a custom jar which will export the traces directly to the database. This can be a new idea to save the traces to database directly eliminating the need of collector.

hi @Yahire3388! this is a very big undertaking, I'd strongly recommend looking at existing solutions (either OSS or commercial) for storing and viewing your traces

**Reply by Yahire3388** [reply]
Hi,

I am trying this [extension jar](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension). But when i run the java agent with this extension jar, it is not adding any of the custom attribute or anything.

When i looked in the gradle script it is something like this :
tasks {
  test {
    useJUnitPlatform()

    inputs.files(layout.files(tasks.shadowJar))
    inputs.files(layout.files(tasks.extendedAgent))

    systemProperty 'io.opentelemetry.smoketest.agentPath', configurations.otel.singleFile.absolutePath
    systemProperty 'io.opentelemetry.smoketest.extendedAgentPath', tasks.extendedAgent.archiveFile.get().asFile.absolutePath
    systemProperty 'io.opentelemetry.smoketest.extensionPath', tasks.shadowJar.archiveFile.get().asFile.absolutePath
  }

  compileJava {
    options.release.set(8)
  }

  assemble.dependsOn(shadowJar)
}

Do we have t add any task regarding main class? Also, I was hoping if i add db connection code in DemoSpanExporter.java it will store the data in DB. Is it a correct way?

public class DemoSpanExporter implements SpanExporter {
  @Override
  public CompletableResultCode export(Collection<SpanData> spans) {
    System.out.printf("%d spans exported%n", spans.size());

**// Add database code here**

    return CompletableResultCode.ofSuccess();
  }

  @Override
  public CompletableResultCode flush() {
    return CompletableResultCode.ofSuccess();
  }

  @Override
  public CompletableResultCode shutdown() {
    return CompletableResultCode.ofSuccess();
  }
}

**Reply by mateuszrzeszutek** [reply]
> Do we have t add any task regarding main class?

No, there's no need to have a main class in an extension -- the extension classes are loaded via the `ServiceLoader` mechanism (look for either `@AutoService` annotation or `META-INF/services` files).

> Also, I was hoping if i add db connection code in DemoSpanExporter.java it will store the data in DB. Is it a correct way?
> ```
> // Add database code here
> ```

Yes, that looks like more or less the correct place.
